packages/flutter_tools/lib/src/ios/code_signing.dart

```diff
@@ -116,229 +159,787 @@ Future<Map<String, String>?> getCodeSigningIdentityDevelopmentTeamBuildSetting({
-  if (signingCertificateId == null) {
+  /// Validates the saved provisioning profile still exists and that there is a
+  /// valid identity/certificate for the profile.
+  ///
+  /// Returns null if profile cannot be found, parsed, or validated.
+  Future<_ProvisioningProfile?> _validateSavedProfile(
+    String savedProfilePath,
+    List<String> validCodeSigningIdentities,
+  ) async {
+    final File savedProfile = _fileSystem.file(savedProfilePath);
+    if (!savedProfile.existsSync()) {
+      _logger.printError('Unable to find saved provisioning profile $savedProfilePath');
+      return null;
+    }
+    final _ProvisioningProfile? parsedProfile = await _parseProvisioningProfile(savedProfile);
+    if (parsedProfile == null) {
+      return null;
+    }
+    for (final File cert in parsedProfile.developerCertificates) {
+      final String? identity = await _validateIdentityFromCert(cert, validCodeSigningIdentities);
+      if (identity != null) {
+        return parsedProfile;
+      }
+    }
+    _logger.printError('Unable to find a valid certificate matching the provisioning profile.');
     return null;
   }
 
-  String signingCertificateStdout;
-  try {
-    signingCertificateStdout =
-        (await processUtils.run(<String>[
-          'security',
-          'find-certificate',
-          '-c',
-          signingCertificateId,
-          '-p',
-        ], throwOnError: true)).stdout.trim();
-  } on ProcessException catch (error) {
-    logger.printTrace("Couldn't find the certificate: $error.");
-    return null;
+  /// Decode and convert a .mobileprovision file to a .plist file and then
+  /// parse the .plist into [_ProvisioningProfile].
+  Future<_ProvisioningProfile?> _parseProvisioningProfile(File provisioningProfileFile) async {
+    final Directory profilesDirectory = _fileSystem.systemTempDirectory.childDirectory(
+      'provisioning_profiles',
+    );
+    profilesDirectory.createSync(recursive: true);
+    final File decodedProfile = profilesDirectory.childFile(
+      'decoded_profile_${provisioningProfileFile.basename}.plist',
+    );
+    try {
+      await _processUtils.run(<String>[
+        'security',
+        'cms',
+        '-D',
+        '-i',
+        provisioningProfileFile.path,
+        '-o',
+        decodedProfile.path,
+      ], throwOnError: true);
+    } on ProcessException catch (error) {
+      _logger.printError('Unexpected failure from security: $error.');
+      return null;
+    }
+    if (!decodedProfile.existsSync()) {
+      _logger.printError('Failed to decode ${provisioningProfileFile.basename}');
+      return null;
+    }
+    final Map<String, Object> contents = _plistParser.parseFile(decodedProfile.path);
+    try {
+      final _ProvisioningProfile parsedProfile = _ProvisioningProfile.fromPlist(
+        provisioningProfileFile.path,
+        contents,
+        fileSystem: _fileSystem,
+      );
+      return parsedProfile;
```

Comment:
Nit: consider returning directly:
```suggestion
      return _ProvisioningProfile.fromPlist(
        provisioningProfileFile.path,
        contents,
        fileSystem: _fileSystem,
      );
```

I don't feel strongly about this

Reply:
That's weird, return directly in a try catch should catch. This seems to work as expected: https://dartpad.dev/?id=6d9943629b1821f6858f50f41c2aad48

Were you doing something else?

Reply:
Okay weird. I swear it wasn't catching when I originally wrote it but now it does 🙈 I even remember being super confused and thinking I learned something new lol. Oh well, looks like it catches properly when returned directly so I updated it.